[CAS] Allow uncached job from CAS based dependency scanning

cachemeifyoucan opened the pull request:
Create a path that swift-frontend can execute an uncached job from modules built with CAS based explicit module build. The new flag `-import-module-from-cas` will allow an uncached build to load module from CAS, and combined with source file from real file system to build the current module. This allows quick iterations that bypasses CAS, without full dependency scanning every time in between.

rdar://152441866

<!--
If this pull request is targeting a release branch, please fill out the
following form:
https://github.com/swiftlang/.github/blob/main/PULL_REQUEST_TEMPLATE/release.md?plain=1

Otherwise, replace this comment with a description of your changes and
rationale. Provide links to external references/discussions if appropriate.
If this pull request resolves any GitHub issues, link them like so:

  Resolves <link to issue>, resolves <link to another issue>.

For more information about linking a pull request to an issue, see:
https://docs.github.com/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue
-->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

cachemeifyoucan commented:
@swift-ci please smoke test

benlangmuir commented on lib/Frontend/Frontend.cpp:
```diff
@@ -640,8 +640,9 @@ bool CompilerInstance::setUpVirtualFileSystemOverlays() {
-    if (Invocation.getCASOptions().HasImmutableFileSystem) {
+  if (CASOpts.EnableCaching) {
+    if (!CASOpts.ClangIncludeTree.empty() ||
+        !CASOpts.ClangIncludeTreeFileList.empty()) {
```
I'm not sure what is going on with this change

For `requireCASFS` changing to `EnableCaching`: what happens if we have `InputFileKey` or `BridgingHeaderPCHCacheKey` but not caching? 

For `HasImmutableFileSystem`: isn't this the same conditions as that flag? What's the reason to change it here?

cachemeifyoucan replied:
> For requireCASFS changing to EnableCaching: what happens if we have InputFileKey or BridgingHeaderPCHCacheKey but not caching?

The entire bag of inputs from CAS are tied with EnableCaching. They are basically all ignored if caching is not enabled.

HasImmutableFileSystem is actually not the same has the check here when it is inside a sub invocation. Maybe HasImmutableFileSystem is wrong and should be turned into a instance method that needs to be computed from current CASOption state.

cachemeifyoucan replied:
The code change is actually no-op, just to make it more readable. If requires cleanup, I will do in a separate commit.

cachemeifyoucan committed:
[CAS] Allow uncached job from CAS based dependency scanning

Create a path that swift-frontend can execute an uncached job from
modules built with CAS based explicit module build. The new flag
-import-module-from-cas will allow an uncached build to load module
from CAS, and combined with source file from real file system to build
the current module. This allows quick iterations that bypasses CAS,
without full dependency scanning every time in between.

rdar://152441866
